Packaging grpc-tools from grpc-node

**Project description**

The https://github.com/grpc/grpc-node repository has a number of GRPC related packages usable in the Node ecosystem.

In particular the https://github.com/grpc/grpc-node/tree/master/packages/grpc-tools tries to bring `grpc_tools_node_protoc` and `grpc_tools_node_protoc_plugin` for generating JS code from protobuf spec.

However it's using C++ code inside, and it's not really built by the `node-pre-gyp`. It only uses that `node-pre-gyp` to download a precompiled binary hosted on node-precompiled-binaries.grpc.io.

So this package doesn't work just be overriding from `node2nix` output.

```
{ pkgs ? import <nixpkgs> {} }:

let p = import ./nix/default.nix { pkgs = pkgs; };
in p.package.override (oldAttrs: {
  buildInputs = oldAttrs.buildInputs ++ [ pkgs.nodePackages.node-pre-gyp ];
})
```

The above does not work, with `nix-build`, and just produces:

```
> node-pre-gyp install

node-pre-gyp WARN Using needle for node-pre-gyp https download
node-pre-gyp ERR! install error
node-pre-gyp ERR! stack Error: getaddrinfo ENOTFOUND node-precompiled-binaries.grpc.io
node-pre-gyp ERR! stack     at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:60:26)
node-pre-gyp ERR! System Linux 4.19.128
node-pre-gyp ERR! command "/nix/store/rcxlrj41836ps5iqv3jikqn6h7msbjpw-nodejs-12.15.0/bin/node" "/nix/store/hzw29a3cf0ync2mn4jbwd1x5sfpqdiwi-node_grpc-tools-1.9.0/lib/node_modules/grpc-tools/node_modules/.bin/node-pre-gyp" "install"
node-pre-gyp ERR! cwd /nix/store/hzw29a3cf0ync2mn4jbwd1x5sfpqdiwi-node_grpc-tools-1.9.0/lib/node_modules/grpc-tools
node-pre-gyp ERR! node -v v12.15.0
node-pre-gyp ERR! node-pre-gyp -v v0.15.0
node-pre-gyp ERR! not ok
getaddrinfo ENOTFOUND node-precompiled-binaries.grpc.io
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! grpc-tools@1.9.0 install: `node-pre-gyp install`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the grpc-tools@1.9.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /build/.npm/_logs/2020-07-15T11_46_32_675Z-debug.log

builder for '/nix/store/avrmmjcpa2yybrdv77n5w95q86dvsvjz-node_grpc-tools-1.9.0.drv' failed with exit code 1
error: build of '/nix/store/avrmmjcpa2yybrdv77n5w95q86dvsvjz-node_grpc-tools-1.9.0.drv' failed
```

So we need some help packaging the C++ in Nix.

**Metadata**

* homepage URL: https://github.com/grpc/grpc-node/tree/master/packages/grpc-tools
* source URL: https://github.com/grpc/grpc-node/tree/master/packages/grpc-tools
* license: APL2
* platforms: unix, linux, darwin, ...

--- a/pkgs/development/tools/misc/grpc-tools/default.nix
+++ b/pkgs/development/tools/misc/grpc-tools/default.nix
@@ -1,0 +1,40 @@
+{ lib
+, stdenv
+, cmake
+, fetchFromGitHub
+}:
+
+stdenv.mkDerivation rec {
+  pname = "grpc-tools";
+  version = "1.10.0";
+
+  src = fetchFromGitHub {
+    owner = "grpc";
+    repo = "grpc-node";
+    rev = "3a094f01711942f79abd8a536c45a91b574d626f"; # version 1.10.0 was not tagged
+    sha256 = "1a7l91kxc3g7mqfqvhc3nb7zy0n21ifs5ck0qqg09qh3f44q04xm";
+    fetchSubmodules = true;
+  };
+
+  sourceRoot = "source/packages/grpc-tools";
+
+  nativeBuildInputs = [ cmake ];
+
+  installPhase = ''
+    install -Dm755 -t $out/bin grpc_node_plugin
+    install -Dm755 -t $out/bin deps/protobuf/protoc
+  '';
+
+  meta = with lib; {
+    description = "Distribution of protoc and the gRPC Node protoc plugin for ease of installation with npm";
+    longDescription = ''
+      This package distributes the Protocol Buffers compiler protoc along with
+      the plugin for generating client and service objects for use with the Node
+      gRPC libraries.
+    '';
+    homepage = "https://github.com/grpc/grpc-node/tree/master/packages/grpc-tools";
+    license = licenses.asl20;
+    platforms = platforms.all;
+    maintainers = [ maintainers.nzhang-zh ];
+  };
+}

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -4789,6 +4789,8 @@ in
 
   grpcui = callPackage ../tools/networking/grpcui { };
 
+  grpc-tools = callPackage ../development/tools/misc/grpc-tools { };
+
   grub = pkgsi686Linux.callPackage ../tools/misc/grub ({
     stdenv = overrideCC stdenv buildPackages.pkgsi686Linux.gcc6;
   } // (config.grub or {}));

